plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/listener/SearchQueryLatencyListener.java

```diff
@@ -0,0 +1,95 @@
+import org.opensearch.common.inject.Inject;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.plugin.insights.core.service.TopQueriesByLatencyService;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Locale;
+
+import static org.opensearch.plugin.insights.settings.QueryInsightsSettings.TOP_N_LATENCY_QUERIES_ENABLED;
+import static org.opensearch.plugin.insights.settings.QueryInsightsSettings.TOP_N_LATENCY_QUERIES_SIZE;
+import static org.opensearch.plugin.insights.settings.QueryInsightsSettings.TOP_N_LATENCY_QUERIES_WINDOW_SIZE;
+
+/**
+ * The listener for top N queries by latency
+ *
+ * @opensearch.internal
+ */
+public final class SearchQueryLatencyListener extends SearchRequestOperationsListener {
+    private static final ToXContent.Params FORMAT_PARAMS = new ToXContent.MapParams(Collections.singletonMap("pretty", "false"));
+
+    private static final Logger log = LogManager.getLogger(SearchQueryLatencyListener.class);
+
+    private final TopQueriesByLatencyService topQueriesByLatencyService;
+
+    @Inject
+    public SearchQueryLatencyListener(ClusterService clusterService, TopQueriesByLatencyService topQueriesByLatencyService) {
+        this.topQueriesByLatencyService = topQueriesByLatencyService;
+        this.setEnabled(clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_ENABLED));
+        this.topQueriesByLatencyService.setTopNSize(clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_SIZE));
+        this.topQueriesByLatencyService.setWindowSize(clusterService.getClusterSettings().get(TOP_N_LATENCY_QUERIES_WINDOW_SIZE));
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(TOP_N_LATENCY_QUERIES_ENABLED, this::setEnabled);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(TOP_N_LATENCY_QUERIES_SIZE, this.topQueriesByLatencyService::setTopNSize);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(TOP_N_LATENCY_QUERIES_WINDOW_SIZE, this.topQueriesByLatencyService::setWindowSize);
+    }
+
+    @Override
+    public void setEnabled(boolean enabled) {
+        super.setEnabled(enabled);
+        this.topQueriesByLatencyService.setEnabled(enabled);
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return super.isEnabled();
+    }
+
+    @Override
+    public void onPhaseStart(SearchPhaseContext context) {}
+
+    @Override
+    public void onPhaseEnd(SearchPhaseContext context, SearchRequestContext searchRequestContext) {}
+
+    @Override
+    public void onPhaseFailure(SearchPhaseContext context) {}
+
+    @Override
+    public void onRequestStart(SearchRequestContext searchRequestContext) {}
+
+    @Override
+    public void onRequestEnd(SearchPhaseContext context, SearchRequestContext searchRequestContext) {
+        SearchRequest request = context.getRequest();
+        try {
+            topQueriesByLatencyService.ingestQueryData(
```

Comment:
If we need to store additional context for each topN query such as query shape, we could pass it here right? And add an additional attribute in `SearchQueryLatencyRecord`

Reply:
yes, there's a map to pass in additional context attributes.